src/main/groovy/io/cify/framework/core/DeviceManager.groovy

```diff
@@ -135,13 +137,74 @@ class DeviceManager implements IDeviceManager {
             Device device = new Device(deviceId, category, desiredCapabilities)
             devices.add(device)
             setDeviceActive(device)
+        } catch (all) {
+            LOG.debug(MARKER, all.message, all)
+            throw new CifyFrameworkException("Failed to create device. Cause: $all.message")
+        }
+
+        addDeviceToTestReport(deviceId, category.toString())
+        capabilities.setAvailable(desiredCapabilities.getCapability(CAPABILITY_ID) as String, false)
+
+        return getActiveDevice(deviceId)
+    }
+
+    /**
+     * Creates device with capabilities identified by capabilityId
+     * @param capabilityId capability identifier
+     * @return Device
+     */
+    @Override
+    Device createDevice(String capabilityId) {
+        createDevice(capabilityId, generateRandomDeviceId())
+    }
+
+    /**
+     * Creates device with capabilities identified by capabilityId and with unique deviceId
+     * @param capabilityId capability identifier
+     * @param deviceId unique device id
+     * @return Device
+     */
+    @Override
+    Device createDevice(String capabilityId, String deviceId) {
+        DeviceCategory category
+        DesiredCapabilities desiredCapabilities
+        LOG.debug(MARKER, "Create new device with capabilityId $capabilityId")
+        try {
+
+            if (capabilityId == null || capabilityId.isEmpty()) {
+                throw new CifyFrameworkException("CapabilityId is null or empty")
+            }
+
+            if (deviceId == null || deviceId.isEmpty()) {
+                throw new CifyFrameworkException("DeviceId is null or empty")
+            }
+
+            if (hasActiveDevice(deviceId)) {
+                throw new CifyFrameworkException("Device with id $deviceId already exists")
+            }
+
+            desiredCapabilities = capabilities.toDesiredCapabilities(capabilityId)
+
+            if (desiredCapabilities.asMap().isEmpty()) {
+                throw new CifyFrameworkException("No capabilities provided for $capabilityId")
+            }
+
+            credentials.each { key, value ->
+                desiredCapabilities.setCapability(key, value)
+            }
 
+            category = capabilities.getCategory(capabilityId)
+            Device device = new Device(deviceId, category, desiredCapabilities)
+            devices.add(device)
+            setDeviceActive(device)
         } catch (all) {
             LOG.debug(MARKER, all.message, all)
-            throw new CifyFrameworkException("Failed to create device cause $all.message")
+            throw new CifyFrameworkException("Failed to create device. Cause: $all.message")
```

Comment:
Maybe it make sense not to throw an exception but return a Null ?

Reply:
Exception seems to be more informative, isn't it?